### PR TITLE
[Guided tour] plan

### PR DESCRIPTION
**Pull request by duggiemitchell**
## Cute animal pic
_Because everyone likes pictures of [animals](https://unsplash.com/t/animals)._
tj

## Link to Trello card (if applicable)
https://trello.com/c/zyv78LfY

TODO:
- ~Remaining content for on boarding steps (epic table name, status, & collaborators column)~
- [x] Icons for walk-through modal
- [x] "?" dropdown to get back to walkthrough
- [x] `PUT` to set `user.onboarded_at` at tour start
- [x] Back-fill Tests

## Description
_The commit messages say what you did; this should explain why and/or how._
- [x] Description is in Trello card

 Walkthrough for new users to introduce tasks and epics.

### CODE REVIEW

**Comment by duggiemitchell** [comment]
@jerivas we need a way to know if a user is a "new user" to show them a tour.  Maybe just a new field on the user modal (i.e. `is_onboarded`?)

**Comment by duggiemitchell** [comment]
[As a new user, I want a walkthrough to introduce tasks and epics. (Plan Walkthrough)](https://trello.com/c/zyv78LfY/260-as-a-new-user-i-want-a-walkthrough-to-introduce-tasks-and-epics-plan-walkthrough)

**Comment by jerivas** [comment]
@duggiemitchell I added a new `onboarded_at` field to the User instances, accessible via `/api/user/` and  the `data-user` attribute of `<main>`. I figured this might be a bit more useful than just a boolean in case you ever need to run the onboarding again (for example, after major changes to the UI). We could also modify it to let you store an arbitrary string or something else, just let me know.

To set the date of the onboarding send a PUT request to `/api/complete_onboarding/`. This works similarly to the `agreed_to_tos` endpoint.

File: src/js/components/projects/detail.tsx
Comment (duggiemitchell) [inline src/js/components/projects/detail.tsx:104]: @jgerigmeyer this is probably the point in which we should 'PUT' to the user endpoint to update" `onboarding_at` ? I'm wondering how we get someone to do onboarding for 'Play' and 'Help'

**Comment by duggiemitchell** [comment]
@jgerigmeyer, is there something different that needs to happen for the ci to install new dependencies?   the build is fails at trying to install `joyride`
Screenshot from 2021-02-15 11-24-47

File: metecho/api/models.py
Comment (duggiemitchell) [inline metecho/api/models.py:82]: @jgerigmeyer do we want to know which onboarding flows a user has completed?  My concern with using a timestamp would be how to know whether to guide someone through the other tours if they've just completed one.

**Comment by jgerigmeyer** [comment]
> @jgerigmeyer, is there something different that needs to happen for the ci to install new dependencies? the build is fails at trying to install `joyride`

@duggiemitchell Looks like you must have installed the `react-joyride` dependency locally somehow (globally?), but it wasn't added to `package.json` or `yarn.lock`. Fixed.

File: src/js/components/projects/detail.tsx
Comment (jgerigmeyer) [reply src/js/components/projects/detail.tsx]: > @jgerigmeyer this is probably the point in which we should 'PUT' to the user endpoint to update" `onboarding_at` ?

@duggiemitchell Actually I think we want to set this initially when the modal opens. Even if a user doesn't complete the walkthrough, we don't want to show them this modal a second time.

> I'm wondering how we get someone to do onboarding for 'Play' and 'Help'

The little `?` dropdown menu needs to be added to the global header, next to the user icon. Users can't get back to the landing modal, but they can directly start any of the walkthroughs via that dropdown.

File: metecho/api/models.py
Comment (jgerigmeyer) [reply metecho/api/models.py:82]: No, we don't want to force users to complete all of the walkthroughs. We just show them the modal once initially, then that's it.

**Review by jgerigmeyer — COMMENTED** [review]
@duggiemitchell Looking good! I did a bit of review/cleanup. A few notes:

- I notice you had various table headers highlighted as steps. I don't see those in the mockup, and I confirmed with Sondra that we only want the one "Create an Epic" steps for this initial walkthrough.
- We still need to add the `?` dropdown menu, so users can access the walkthrough again.
- Related to that, I posted a comment in the mockup (https://xd.adobe.com/view/f58dda9a-72c4-4b74-8b31-3e6ddc4334e4-a0de/) for @SondraE... The mockup shows a static image of the `?` dropdown in the modal itself. I'd propose not adding that, but instead highlighting the `?` dropdown as the final step of every walkthrough.

File: test/js/components/projects/detail.test.jsx
Comment (jgerigmeyer) [inline test/js/components/projects/detail.test.jsx:321]: It often works either way, but `fireEvent` returns a promise so it's probably best-practice to always use `await` with it.

File: src/js/components/tour/plan.tsx
Comment (jgerigmeyer) [inline src/js/components/tour/plan.tsx:45]: Like I said in my review comments yesterday, I think we want this "onboard" action to run immediately when the landing modal initially opens, not after the onboarding is done.

File: src/js/components/projects/detail.tsx
Comment (duggiemitchell) [inline src/js/components/projects/detail.tsx:80]: @jgerigmeyer these are the code branches I mentioned in standup that aren't covered in the tests; could be why the tests for header are not passing as well.

File: src/js/components/header.tsx
Comment (duggiemitchell) [inline src/js/components/header.tsx:46]: @jgerigmeyer also meant to mention, these icons aren't showing up; passing with `category` may not be needed, but makes we wonder if we do need to use the [listItemRenderer](https://react.lightningdesignsystem.com/components/menu-dropdowns/#prop-listItemRenderer) prop instead?

**Review by stacyk — COMMENTED** [review]
I added some styles and edited the markup a bit to fit LDS guidelines ... but is there anything else for me to do here? The fade-in wasn't working on the SLDS styles... maybe because .1 is too fast to see while the page loads and it is already fully visible by then so there isn't a visible fade. That is why I added an animation to slow this down a bit.

I played around with the icons... if I remove category the icon prints in the html, but it is inside of the span so something there wasn't quite right.

File: locales/en/translation.json
Comment (duggiemitchell) [reply locales/en/translation.json:45]: I think we can defer to @SondraE ? I'm fine either way.

File: test/js/components/projects/detail.test.jsx
Comment (duggiemitchell) [reply test/js/components/projects/detail.test.jsx:321]: oops, was listening to vs code warnings saying `await` would have no affect.

File: test/js/components/projects/detail.test.jsx
Comment (duggiemitchell) [inline test/js/components/projects/detail.test.jsx:321]:
```suggestion
     await fireEvent.click(getByTitle('Close'));
```

File: test/js/components/projects/detail.test.jsx
Comment (jgerigmeyer) [reply test/js/components/projects/detail.test.jsx:321]: Interesting -- I've never seen that warning, and I'm not sure how VSC would know whether it has an effect 🤔

File: locales/en/translation.json
Comment (SondraE) [reply locales/en/translation.json:45]: @stacyk I go back and forth on "select" or "click". We can definitely go with Select.

File: locales/en/translation.json
Comment (SondraE) [reply locales/en/translation.json:361]: @stacyk Great question. As I've been writing all the text for the walkthroughs and self-guided tour, I've been capitalizing the important nouns/objects (Project, Epic, Task, Org, Collaborator). It's useful to me, but I'm curious if you think others will find that useful or annoying. If useful, then yes, I'd like to be consistent with those uppercase letters across the various tutorials.

File: locales/en/translation.json
Comment (SondraE) [reply locales/en/translation.json:362]: @stacyk I love "turn-on the self-guided tour"! That is more clear. I'd love for you to review all my text for the tour and walkthroughs. Almost ready for review...

File: locales_dev/en/translation.json
Comment (SondraE) [reply locales_dev/en/translation.json:357]: @stacyk See above comment about capitalizing.

File: test/js/components/projects/detail.test.jsx
Comment (jgerigmeyer) [reply test/js/components/projects/detail.test.jsx]: @duggiemitchell You're right -- I was getting confused with the Vue version, where `fireEvent` does return a promise. These aren't needed here 👍

**Comment by jgerigmeyer** [comment]
@stacyk I think this is functionally complete. I need to backfill some tests, but in the meantime I notice since I added a second step to the walkthrough, the "Back" btn could probably use some contrast cleanup.

Screen Shot 2021-03-24 at 5 54 17 PM

**Review by jgerigmeyer — APPROVED** [review]

**Comment by jgerigmeyer** [comment]
[SVGs when used inline with the data: encoding aren't showing up in Safari](https://trello.com/c/qLjodABH/272-svgs-when-used-inline-with-the-data-encoding-arent-showing-up-in-safari)

File: src/js/components/tour/plan.tsx
Comment (stacyk) [inline src/js/components/tour/plan.tsx:118]: I'm making some changes I haven't pushed yet to the buttons, but I wonder if you could help @jgerigmeyer? Is there a way to edit the buttonBase instead of each of the 4 button types? 
https://github.com/gilbarbara/react-joyride/blob/c33746d542fd8203d2e85ed07a0c07753c3cc86f/src/styles.js#L15

When I add buttonBase inside the styles={{ }}  here it turns a different color and underlines everything in red. I assume I can't add this base style within this styles block... but how is it added then?

Also, These styles will be used on all of the tours, so should they be moved into a different file so they arent duplicated in the play and help walk throughs as well?

Or, is there a way to add a lighting design class to each of the elements so we don't have to override the styles? Or is it too tricky to access the markup like that? I don't mind which way we go, but I assume this style duplication will be a bummer to maintain if there are changes down the road.

Comment (jgerigmeyer) [reply src/js/components/tour/plan.tsx:118]: > I'm making some changes I haven't pushed yet to the buttons, but I wonder if you could help @jgerigmeyer? Is there a way to edit the buttonBase instead of each of the 4 button types?
> https://github.com/gilbarbara/react-joyride/blob/c33746d542fd8203d2e85ed07a0c07753c3cc86f/src/styles.js#L15
> 
> When I add buttonBase inside the styles={{ }} here it turns a different color and underlines everything in red. I assume I can't add this base style within this styles block... but how is it added then?

@stacyk 

No, we can't directly override the `buttonBase` to style all buttons at once. That's only defined internally in react-joyride, and then the default Joyride styles override those individually, and then our styles override those individually as well. So we'd have to style each of them separately.

> Also, These styles will be used on all of the tours, so should they be moved into a different file so they arent duplicated in the play and help walk throughs as well?

Sure, I think you're 💯% right... but this PR is just for the first walkthrough, so I'd say let's worry about abstracting patterns once we have patterns to abstract. 😄 

> Or, is there a way to add a lighting design class to each of the elements so we don't have to override the styles? Or is it too tricky to access the markup like that?

I don't think so, no. I mean, we can choose to override _all_ the markup if we want -- but I'd rather not go that route unless we really want to. We can't just attach classes, without rolling our own markup entirely. And this seems pretty close to me...

> I don't mind which way we go, but I assume this style duplication will be a bummer to maintain if there are changes down the road.

Yeah, I get that. I'd lean toward creating our own style definition here (using their markup), then when we add the next walkthrough we'll move this to its own file and use it for all walkthroughs. Does that work?

**Review by stacyk — APPROVED** [review]
Not perfect, but we are using Joyride's style options where possible, along with some CSS to override the inline styles to add some state variation on hover and focus to match LDS popover walkthroughs.

**Comment by jgerigmeyer** [comment]
Looks good! Requesting review from @mirisuzanne (Sass) and @duggiemitchell (JS).

File: src/js/components/tour/plan.tsx
Comment (duggiemitchell) [inline src/js/components/tour/plan.tsx:86]: **suggestion**: define these styles outside of this file for reuse other tours.

(Probably not a blocker here, but we may want this abstracted in future tour stories)

File: test/js/components/epics/detail.test.jsx
Comment (duggiemitchell) [inline test/js/components/epics/detail.test.jsx:1]: neat 🎊

**Review by duggiemitchell — COMMENTED** [review]
I can't approve but LGTM! 
giphy

File: src/js/components/tour/guided.tsx
Comment (jgerigmeyer) [inline src/js/components/tour/guided.tsx:23]: @dvdherron We basically use `useCallback` any time a function is passed as an argument to another React component. At a high level, it basically ensures that the function does not change unless one of its dependencies changes -- that way we're not triggering unnecessary re-renders of the child component.

https://reactjs.org/docs/hooks-reference.html#usecallback

File: locales_dev/en/translation.json
Comment (jgerigmeyer) [inline locales_dev/en/translation.json:34]: @dvdherron Changes to this file come automatically as part of the build process, to help see when there are user-facing strings (which require translation) that have changed during development. But then like @duggiemitchell said, before merging we manually copy over (and fix, if needed) the translations to `locales/en/translation.json`, which is the translation file actually used in production.

File: src/js/components/header.tsx
Comment (mirisuzanne) [inline src/js/components/header.tsx:43]: Probably not new in this branch, but it looks like icons are showing up as unlabeled `role=diagram` as the default role for svgs. We probably want to add `aria-hidden="true" focusable="false"` like we do on other sites, if that's an option.

File: src/js/components/tour/landing.tsx
Comment (mirisuzanne) [inline src/js/components/tour/landing.tsx:59]: Dialogs should have labels: https://developer.mozilla.org/en-US/docs/Web/Accessibility/Understanding_WCAG/Text_labels_and_names#dialogs_should_be_labeled

(I saw that warning when inspecting the joyride dialogs, but assume it applies here as well - and I don't see where to note it in the joyride code)

File: src/js/components/tour/styles.tsx
Comment (mirisuzanne) [inline src/js/components/tour/styles.tsx:16]: Do we need to use px values for fonts? Worried this won't respond well to font preferences.

File: src/sass/components/_tour.scss
Comment (mirisuzanne) [inline src/sass/components/_tour.scss:7]: Wouldn't it be cool if we had CSS layers?

File: static/images/backpack-sm.svg
Comment (mirisuzanne) [inline static/images/backpack-sm.svg:1]: these svgs could probably also get hidden from aria/focus?

**Review by mirisuzanne — APPROVED** [review]
functionality looks great. Left a few notes on a11y – I think the modal label is probably most important of those, if we're able to do something about it.

File: src/sass/components/_tour.scss
Comment (stacyk) [reply src/sass/components/_tour.scss:7]: It would have come in really handy.

File: src/js/components/header.tsx
Comment (jgerigmeyer) [reply src/js/components/header.tsx:43]: @mirisuzanne I might be misunderstanding, but from what I can tell all of our SVG icons have `aria-hidden="true"` as per SLDS specs. I think we can add `focusable="false"` if that's important.

File: src/js/components/tour/landing.tsx
Comment (jgerigmeyer) [reply src/js/components/tour/landing.tsx:59]: @mirisuzanne We don't really control this markup, but I'm seeing both `aria-label` and `aria-describedby`. Is there something else missing?

Screen Shot 2021-03-29 at 10 49 00 AM

File: static/images/backpack-sm.svg
Comment (jgerigmeyer) [reply static/images/backpack-sm.svg:1]: @mirisuzanne I think these are already getting `aria-hidden="true"`?

Screen Shot 2021-03-29 at 10 52 03 AM

Comment (mirisuzanne) [reply static/images/backpack-sm.svg:1]: oh, ok - not sure how I missed that.

**Pull request merged**
